Sourcery/Parsing/FileParser.swift

```diff
@@ -307,6 +312,73 @@ extension FileParser {
         return true
     }
 
+    fileprivate func extractGenericsDeclaration(source: String) -> String {
+        var declaration = source
+        if let whereStart = declaration.index(of: "where") {
+            declaration = String(declaration[declaration.startIndex ..< whereStart])
+        }
+
+        if let start = source.index(of: "<") {
+            var end = start
+            var openingAngles = 1
+            while openingAngles > 0 && end != source.endIndex {
+                end = source.index(after: end)
+                if source[end] == ">" {
+                    openingAngles -= 1
+                } else if source[end] == "<" {
+                    openingAngles += 1
+                }
+            }
+
+            return String(source[source.index(after: start) ..< end])
+        }
+        return ""
+    }
+
+    fileprivate func extractGenericsWhereClause(source: String) -> String {
+        if let start = source.index(of: "where") {
+            return String(source[source.index(start, offsetBy: "where".count) ..< source.endIndex])
+        }
+        return ""
+    }
+
+    fileprivate func parseGenerics(declaration: String, whereClause: String) -> [Generic] {
+        func extractGenerics(source: String) -> [Generic] {
+            return source.components(separatedBy: ",")
```

Comment:
we should use `commaSeparated` method, it will ensure that separators are processed correctly, otherwise things like `Foo<A: [(String, String)]>` will be parsed incorrectly